docs(delivery): added first draft of e2e ack guide (#11760)
vectordotdev/vector: website/content/en/guides/level-up/e2e-ack-guide.md

---
title: Ensure data is always successfully delivered
short: End-to-end acknowledgement
description: Learn how to use end-to-end acknowledgement
author_github: https://github.com/barieom
domain: delivery

> weight: 5

Reviewer: <img width="787" alt="image" src="https://user-images.githubusercontent.com/11903873/157560843-cc58581d-38a7-4511-937b-f8147a29accb.png">

This might need to be changed, or the others. I assume that's driving the guide numbering

tags: ["delivery", "logs", "level up", "guides", "guide"]
---

{{< requirement >}}
Before you begin, this guide assumes the following:

* You understand the [basic Vector concepts][concepts]
* You understand [how to set up a basic pipeline][pipeline]

[concepts]: /docs/about/concepts
[pipeline]: /docs/setup/quickstart
{{< /requirement >}}

Vector is often deployed in mission critical environments where loss of data is
non-negotiable. Ergo, assurance that every data flowing through your Vector
pipeline is successfully delivered to the end destination is critical.

We're excited to walk you through exactly you can accomplish this by using
Vector's end-to-end acknowledgement feature.

## Getting Started

You can set and control the acknowledgement feature either at the global level or

> the source level. Let's start with the global configuration option first, as it is

Reviewer: I thought we swapped this to be enabled at the `sink` side

Reviewer 2: Yeah, that is correct, acknowledgements are configured on sinks now instead of sources. When acknowledgements are enabled for a sink, all associated sources that support acknowledgements will wait for delivery. The ones that don't support it, will output a warning message (we should call that out so users know what to look for).

easy as flipping on a switch. As you would expect from a global configuration, the
configuration below turns end-to-end acknowledgement on for every source:

```toml
acknowledgement = true
```

Even if you have a relatively complex topology and sending data from one source to
multiple sinks, by enabling this global config, all supported sources will wait for
acknowledgement from all the sinks. A source that does acknowledgements will wait
forever for an ack before responding

But you can enable acknowledgements at each source level if you want more granular
control over how acknowledgements for specific cases (e.g. you're buffering your
data). You can set acknowledgements at the source and sink level by:

```toml

> [sources.cool_source_id]

Reviewer: `cool_id` 😆

   # Enable or disable waiting for acknowledgements for this sink.
   # Defaults to the global value of `acknowledgements`

>    acknowledgement = true

Reviewer: Same as above, I thought this was moved to just the `sink`s

...

[sinks.cool_source_id]
   acknowledgement = true
```

## Edge cases 

Unsurprisingly, there are a few exceptions and edge cases for the end-to-end
acknowledgement feature. First, as alluded to earlier, not all sources and sinks
are supported because some sources and sinks are unable to provide acknowledgements

> at the protocol level. A list of sources and sinks that are not supported are:

Reviewer: Why did we end up listing components that _don't_ support e2e, rather than listing components that do support it. I'm not sure I feel strongly either way, but I do feel _something_ about it.

easy as flipping on a switch. As you would expect from a global configuration, the
configuration below turns end-to-end acknowledgement on for every source:

```toml
acknowledgement = true
```

Even if you have a relatively complex topology and sending data from one source to
multiple sinks, by enabling this global config, all supported sources will wait for
acknowledgement from all the sinks. A source that does acknowledgements will wait
forever for an ack before responding

But you can enable acknowledgements at each sink level if you want more granular
control over how acknowledgements for specific cases (e.g. you're buffering your
data). You can set acknowledgements at the sink level by doing the following:

```toml
[sinks.very_cool_id]
   # Enable or disable waiting for acknowledgements for this sink.
   # Defaults to the global value of `acknowledgements`
   acknowledgement = true
```

## Edge cases

Unsurprisingly, there are a few exceptions and edge cases for the end-to-end
acknowledgement feature. First, as alluded to earlier, not all sources and sinks
are supported because some sources and sinks are unable to provide acknowledgements

> at the protocol level. A list of sources and sinks that _are_ supported are:

Reviewer 2: This list is likely to be out of date quickly. I'd just mention that you can see which components support acknowledgements on each of their component documentation pages by looking for the `acknowledgments` badge.

- `datadog_events.cue`
- `datadog_logs.cue`
- `datadog_metrics.cue`
- `elasticsearch.cue`
- `file.cue`
- `gcp_cloud_storage.cue`
- `gcp_pubsub.cue`
- `gcp_stackdriver_logs.cue`
- `gcp_stackdriver_metrics.cue`
- `honeycomb.cue`
- `http.cue`
- `humio.cue`
- `influxdb_logs.cue`
- `influxdb_metrics.cue`
- `kafka.cue`
- `logdna.cue`
- `loki.cue`
- `new_relic.cue`
- `new_relic_logs.cue`
- `redis.cue`
- `sematext_logs.cue`
- `sematext_metrics.cue`
- `splunk_hec_logs.cue`
- `splunk_hec_metrics.cue`
- `vector.cue`

Second, when buffering your observability data, the behavior of end-to-end
acknowledgement is a little different. When an event is persisted in a buffer, the
event will be marked as acknowledged to the source.

Third, the final edge case is if you are using any user-space transforms, such
as Lua. Because user-space transforms can drop, combine, split, or create events
that may have no relation to the source event, tracking whether an event has been
successfully delivered gets extremely complex. As such, end-to-end acknowledgement
behavior will be put in the hands of script writers.

## Parting thoughts

This guide intended to serve as a basic guide on how you can start

> leveraging `acknowledgement`. If you have any feedback regarding this feature,

Reviewer 2: ```suggestion
leveraging Vector's end-to-end acknowledgement feature. If you have any feedback regarding this feature,
```

description: Learn how to use end-to-end acknowledgement
author_github: https://github.com/barieom
domain: delivery
weight: 7
tags: ["delivery", "logs", "level up", "guides", "guide"]
---

{{< requirement >}}
Before you begin, this guide assumes the following:

* You understand the [basic Vector concepts][concepts]
* You understand [how to set up a basic pipeline][pipeline]

[concepts]: /docs/about/concepts
[pipeline]: /docs/setup/quickstart
{{< /requirement >}}

Vector is often deployed in mission critical environments where loss of data is
non-negotiable. Ergo, assurance that every data flowing through your Vector
pipeline is successfully delivered to the end destination is critical.

We're excited to walk you through exactly you can accomplish this by using
Vector's end-to-end acknowledgement feature.

## Getting Started

You can set and control the acknowledgement feature either at the global level or
the sink level. Let's start with the global configuration option first, as it is
easy as flipping on a switch. As you would expect from a global configuration, the
configuration below turns end-to-end acknowledgement on for every source:

```toml

> acknowledgement = true

Reviewer 3: Note, `acknowledgement` should be `acknowledgements` everywhere, and the configuration has an `enabled` field. Setting just `acknowledgements = true` works through a compatibility hack.
```suggestion
acknowledgements.enabled = true
```

```

Even if you have a relatively complex topology and sending data from one source to

> multiple sinks, by enabling this global config, all supported sources will wait for
> acknowledgement from all the sinks. A source that does acknowledgements will wait
> forever for an ack before responding.

Reviewer 3: I'm not sure the latter phrase adds anything to this section. This suggestion is a little awkwardly worded, but comes closer to describing the process.
```suggestion
multiple sinks, by enabling this global config, all sources with acknowledgement support will wait before responding for
acknowledgement from all the sinks that its events flow through.
```

> But you can enable acknowledgements at each sink level if you want more granular

Reviewer 3: ```suggestion
But you can enable acknowledgements individually for each sink if you want more granular
```

control over how acknowledgements for specific cases (e.g. you're buffering your

> data). When you enable `acknowledgement` for a supported sink, all associated sources

Reviewer 3: ```suggestion
data). When you enable `acknowledgements` for a supported sink, all sources connected
```

to that sink will wait for delivery. You can set acknowledgements at the sink level by
doing the following:

```toml
[sinks.very_cool_id]
   # Enable or disable waiting for acknowledgements for this sink.
   # Defaults to the global value of `acknowledgements`

>    acknowledgement = true

Reviewer 3: As above:
```suggestion
   acknowledgements.enabled = true
```

```

## Edge cases

Unsurprisingly, there are a few exceptions and edge cases for the end-to-end
acknowledgement feature. First, as alluded to earlier, not all sources and sinks
are supported because some sources and sinks are unable to provide acknowledgements

> at the protocol level. Sinks and sources that don't support acknowledgements will

Reviewer 3: Also, because we haven't implemented support for some yet that do have protocol support.

Author: addressed

> output a warning message to let you know if you try to enable acknowledgements. You

Reviewer 3: Technically, enabling acknowledgements on a sink that don't support acknowledgements will prevent the configuration from loading (ie fatal error). I don't know if you want to make that distinction.

Author: addressed

can find which components support acknowledgements on each of their documentation
pages by looking for the acknowledgments badge.

Second, when buffering your observability data, the behavior of end-to-end

> acknowledgement is a little different. When an event is persisted in a buffer, the

Reviewer 3: Is it worth distinguishing between disk and memory buffers? This behavior is only for disk buffers and I'd hate to cause confusion here.

Author: addressed

---
title: Ensure data is always successfully delivered
short: End-to-end acknowledgement
description: Learn how to use end-to-end acknowledgement
author_github: https://github.com/barieom
domain: delivery
weight: 7
tags: ["delivery", "logs", "level up", "guides", "guide"]
---

{{< requirement >}}
Before you begin, this guide assumes the following:

* You understand the [basic Vector concepts][concepts]
* You understand [how to set up a basic pipeline][pipeline]

[concepts]: /docs/about/concepts
[pipeline]: /docs/setup/quickstart
{{< /requirement >}}

Vector is often deployed in mission critical environments where loss of data is
non-negotiable. Ergo, assurance that every data flowing through your Vector
pipeline is successfully delivered to the end destination is critical.

We're excited to walk you through exactly you can accomplish this by using
Vector's end-to-end acknowledgement feature.

## Getting Started

You can set and control the acknowledgement feature either at the global level or
the sink level. Let's start with the global configuration option first, as it is
easy as flipping on a switch. As you would expect from a global configuration, the

> configuration below turns end-to-end acknowledgement on for every source:

Reviewer 2: ```suggestion
configuration below turns end-to-end acknowledgement on for every sink:
```

Reviewer 3: Well, technically it is turning it on for every source, which is where the behavior change actually happens. This is a point of confusion.

```toml
acknowledgements.enabled = true
```

Even if you have a relatively complex topology and sending data from one source to
multiple sinks, by enabling this global config, all sources with acknowledgement
support will wait before responding for acknowledgement from all the sinks that
its events flow through.

If any sink rejects a message, that failure marks the batch that the message is
part of as _failed_, and the source will respond accordingly. For example, HTTP sources

> will produce some 400 error code, while sources like Kafka that don't handle
> rejection behaves similarly to a positive acknowledgement.

Reviewer 3: The phrase "don't handle rejection" sounds awkward.
```suggestion
will produce some 400 error code, while sources like Kafka that don't have protocol support for
rejecting messages behaves similarly to a positive acknowledgement.
```

> But you can enable acknowledgements individually for each sink if you want more granular

Reviewer 3: Should the order of these two paragraphs be reversed?

```toml
[sinks.very_cool_id]
   # Enable or disable waiting for acknowledgements for this sink.
   # Defaults to the global value of `acknowledgements`
   acknowledgements.enabled = true
```

To note, if any sink rejects a message, that failure marks the batch that the message is

> part of as _failed_, and the source will respond accordingly. For example, HTTP sources
> will produce some 400 error code, while sources like Kafka that don't have protocol
> support for rejecting messages behaves similarly to a positive acknowledgement.

Reviewer 2: ```suggestion
part of as _failed_, and the source will respond accordingly. For example, HTTP sources
will produce a 400 or 500-level error code, while sources like Kafka that don't have protocol
support for rejecting messages will simply not ack the messages, allowing them to be reprocessed.
```

## Edge cases

Unsurprisingly, there are a few exceptions and edge cases for the end-to-end
acknowledgement feature.

First, as alluded to earlier, not all sources and sinks are supported. We are, of
course, working to add support for more sources and sinks, but support for some

> sources and sinks will always be unable because they cannot provide acknowledgements

Reviewer 3: "support … will always be unable" is awkward.

at the protocol level so we'll be adding that overtime.

> Sources that don't support acknowledgements will output a
> warning message to let you know if you try to enable acknowledgements, while sinks

Reviewer 2: ```suggestion
Sources that don't support acknowledgements will output a
warning message to let you know if you try to enable acknowledgements on a connected sink, while sinks
```

that don't support acknowledgements will trigger a fatal error and prevent the
configuration from loading. You can find which components support acknowledgements
on each of their documentation pages by looking for the acknowledgments badge.

Second, when buffering your observability data in a disk, the behavior of end-to-end
acknowledgement is a little different. When an event is persisted in a disk buffer,

> the event will be marked as acknowledged to the source.

Reviewer 3: Is it worth giving a reason why we do this? ie these events will continue be retried in the event of a crash, so we have accepted responsibility, etc.

Third, the final edge case is if you are using any user-space transforms, such
as Lua. Because user-space transforms can drop, combine, split, or create events
that may have no relation to the source event, tracking whether an event has been
successfully delivered gets extremely complex. As such, end-to-end acknowledgement

> behavior will be put in the hands of script writers.

Reviewer 3: Actually, unless I'm completely misremembering what I did, the lua transform works much like the disk buffers here, and no acknowledgements are possible.
